[SPARK-14132][SPARK-14133][SQL] Alter table partition DDLs

andrewor14 opened the pull request:
## What changes were proposed in this pull request?

This implements a few alter table partition commands using the `SessionCatalog`. In particular:

```
ALTER TABLE ... ADD PARTITION ...
ALTER TABLE ... DROP PARTITION ...
ALTER TABLE ... RENAME PARTITION ... TO ...
```

The following operations are not supported, and an `AnalysisException` with a helpful error message will be thrown if the user tries to use them:

```
ALTER TABLE ... EXCHANGE PARTITION ...
ALTER TABLE ... ARCHIVE PARTITION ...
ALTER TABLE ... UNARCHIVE PARTITION ...
ALTER TABLE ... TOUCH ...
ALTER TABLE ... COMPACT ...
ALTER TABLE ... CONCATENATE
MSCK REPAIR TABLE ...
```
## How was this patch tested?

`DDLSuite`, `DDLCommandSuite` and `HiveDDLCommandSuite`

Commit:
Throw exception for MSCK REPAIR TABLE

Commit:
Throw exception on (un)archive partition

Commit:
Implement add partitions

Commit:
Throw exception for exchange partition

Commit:
Implement drop partitions

Commit:
Throw exception for touch/compact/concatenate

Commit:
Implement rename partition

Commit:
Fix style

Commit:
Merge branch 'master' of github.com:apache/spark into alter-partition-ddl

SparkQA commented:
**[Test build #55158 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55158/consoleFull)** for PR 12220 at commit [`eedb529`](https://github.com/apache/spark/commit/eedb5297ce2736e75c3181efeadaa5d6c09e54f1).

andrewor14 commented:
@yhuai

SparkQA commented:
**[Test build #55158 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55158/consoleFull)** for PR 12220 at commit [`eedb529`](https://github.com/apache/spark/commit/eedb5297ce2736e75c3181efeadaa5d6c09e54f1).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `final class VectorizedParquetInputFormat extends ParquetInputFormat[InternalRow] `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55158/
Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into alter-partition-ddl

Commit:
Fix tests

It seems that Hive supports dropping partitions based on partial
specs, where not all partitioned columns are accompanied with
specified values in the spec. Ironically there's no API to
achieve this using the Hive client, so we need to implement it
ourselves in Spark (see HiveClientImpl.scala).

Additionally two tests in HiveCompatibilitySuite use features
that we explicitly do not allow, so those tests are now added
to the blacklist.

SparkQA commented:
**[Test build #55245 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55245/consoleFull)** for PR 12220 at commit [`85c8b8f`](https://github.com/apache/spark/commit/85c8b8f8fda68bcfd1f551e952040d281a5ca38c).

SparkQA commented:
**[Test build #55245 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55245/consoleFull)** for PR 12220 at commit [`85c8b8f`](https://github.com/apache/spark/commit/85c8b8f8fda68bcfd1f551e952040d281a5ca38c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55245/
Test FAILed.

Commit:
Fix tests

SparkQA commented:
**[Test build #55274 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55274/consoleFull)** for PR 12220 at commit [`2a37c75`](https://github.com/apache/spark/commit/2a37c75fe19fd985ec386455570d69b922e17ac8).

SparkQA commented:
**[Test build #55274 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55274/consoleFull)** for PR 12220 at commit [`2a37c75`](https://github.com/apache/spark/commit/2a37c75fe19fd985ec386455570d69b922e17ac8).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55274/
Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into alter-partition-ddl

Commit:
Fix compile after rebase

Commit:
Correct drop partitions purge behavior

and also push ignoreIfNotExists into HiveClient.

SparkQA commented:
**[Test build #55286 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55286/consoleFull)** for PR 12220 at commit [`220141d`](https://github.com/apache/spark/commit/220141dc748d8278513a7fece270f26be3f33717).

SparkQA commented:
**[Test build #55286 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55286/consoleFull)** for PR 12220 at commit [`220141d`](https://github.com/apache/spark/commit/220141dc748d8278513a7fece270f26be3f33717).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55286/
Test PASSed.

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:
```diff
@@ -333,23 +367,32 @@ case class AlterTableExchangePartition(
+ * The syntax of this command is:
+ * {{{
+ *   ALTER TABLE table DROP [IF EXISTS] PARTITION spec1[, PARTITION spec2, ...] [PURGE];
+ *   ALTER VIEW view DROP [IF EXISTS] PARTITION spec1[, PARTITION spec2, ...];
```
I think we do not support `ALTER VIEW view DROP [IF EXISTS] PARTITION spec1[, PARTITION spec2, ...];` because the semantic of having partitions defined with a view is not defined?

yhuai commented on sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala:
```diff
@@ -307,24 +308,57 @@ case class AlterTableSerDeProperties(
+ * The syntax of this command is:
+ * {{{
+ *   ALTER TABLE table ADD [IF NOT EXISTS] PARTITION spec [LOCATION 'loc1']
+ *   ALTER VIEW view ADD [IF NOT EXISTS] PARTITION spec
```
Let's also remove `ALTER VIEW view ADD [IF NOT EXISTS] PARTITION spec`.

Commit:
Views may not be partitioned

A previous patch already throws an exception for us when this
happens. I changed the exception to be AnalysisException so we
can write the test easier. Eventually all of these exceptions
will be made consistent.

This is mainly just a documentation + test change.

andrewor14 commented:
Looks like #12169 already added the check in the parser so we don't have to add it ourselves. The last commit just removes the bad comment and adds two tests for the views.

Commit:
Merge branch 'master' of github.com:apache/spark into alter-partition-ddl

SparkQA commented:
**[Test build #55390 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55390/consoleFull)** for PR 12220 at commit [`3fb9a87`](https://github.com/apache/spark/commit/3fb9a87cc786b23c5bad6a651e4163afbae0a045).

SparkQA commented:
**[Test build #55391 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55391/consoleFull)** for PR 12220 at commit [`12cf973`](https://github.com/apache/spark/commit/12cf973da405d6209f0af6842613cf546bcf8471).

yhuai commented:
LGTM pending jenkins

yhuai commented:
Actually, I have another question. If the table is an external table, we should not delete the data when we drop the partition. What do we do for this case?

yhuai commented:
OK. I think metastore will not delete the data for external tables even if deleteData is set to true (https://github.com/apache/hive/blob/release-0.13.1/metastore/src/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java#L2239-L2251). Do we have tests for external tables? If not, can we add tests? Thanks!

andrewor14 commented:
OK, let's do that in a separate patch.

SparkQA commented:
**[Test build #55390 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55390/consoleFull)** for PR 12220 at commit [`3fb9a87`](https://github.com/apache/spark/commit/3fb9a87cc786b23c5bad6a651e4163afbae0a045).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55390/
Test FAILed.

yhuai commented:
test this please

SparkQA commented:
**[Test build #55396 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55396/consoleFull)** for PR 12220 at commit [`12cf973`](https://github.com/apache/spark/commit/12cf973da405d6209f0af6842613cf546bcf8471).

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -367,9 +367,27 @@ private[hive] class HiveClientImpl(
+        val dropOptions = new PartitionDropOptions
+        dropOptions.ifExists = ignoreIfNotExists
+        dropOptions.purgeData = purge
+        client.dropPartition(db, table, hivePartition.getValues, dropOptions)
```
Oh, just realized that PURGE was added in Hive 0.14. I think we need to either not support it or add tests to VersionsSuite to make sure it works.

andrewor14 replied:
removed

SparkQA commented:
**[Test build #55391 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55391/consoleFull)** for PR 12220 at commit [`12cf973`](https://github.com/apache/spark/commit/12cf973da405d6209f0af6842613cf546bcf8471).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RandomForestClassificationModel(TreeEnsembleModels, JavaMLWritable, JavaMLReadable):`
  * `class HasVarianceCol(Params):`
  * `class RandomForestRegressionModel(TreeEnsembleModels, JavaMLWritable, JavaMLReadable):`
  * `case class DeserializeToObject(`
  * `case class SerializeFromObject(`
  * `public abstract class ColumnVector implements AutoCloseable `
  * `case class DeserializeToObject(`
  * `case class SerializeFromObject(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55391/
Test PASSed.

SparkQA commented:
**[Test build #55396 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55396/consoleFull)** for PR 12220 at commit [`12cf973`](https://github.com/apache/spark/commit/12cf973da405d6209f0af6842613cf546bcf8471).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class RandomForestClassificationModel(TreeEnsembleModels, JavaMLWritable, JavaMLReadable):`
  * `class HasVarianceCol(Params):`
  * `class RandomForestRegressionModel(TreeEnsembleModels, JavaMLWritable, JavaMLReadable):`
  * `case class DeserializeToObject(`
  * `case class SerializeFromObject(`
  * `public abstract class ColumnVector implements AutoCloseable `
  * `case class DeserializeToObject(`
  * `case class SerializeFromObject(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55396/
Test PASSed.

gatorsmile commented:
@andrewor14 I know you are busy. Feel free to let me know if you need me to submit test cases to verify if the data is not deleted when we drop the partition of an external table.

Commit:
Merge branch 'master' of github.com:apache/spark into alter-partition-ddl

Commit:
Don't purge

andrewor14 commented:
@gatorsmile that would be great, thanks!

SparkQA commented:
**[Test build #55565 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55565/consoleFull)** for PR 12220 at commit [`e3f08b3`](https://github.com/apache/spark/commit/e3f08b35efa502f34aabcbba266350e4bfefb0c9).

yhuai commented:
LGTM!

SparkQA commented:
**[Test build #55565 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55565/consoleFull)** for PR 12220 at commit [`e3f08b3`](https://github.com/apache/spark/commit/e3f08b35efa502f34aabcbba266350e4bfefb0c9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55565/
Test PASSed.

yhuai commented:
Merging to master!

gatorsmile commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala:
```diff
@@ -367,9 +367,25 @@ private[hive] class HiveClientImpl(
+      // The provided spec here can be a partial spec, i.e. it will match all partitions
+      // whose specs are supersets of this partial spec. E.g. If a table has partitions
+      // (b='1', c='1') and (b='1', c='2'), a partial spec of (b='1') will match both.
+      val matchingParts = client.getPartitions(hiveTable, s.asJava).asScala
```
I am afraid there is a bug in `getPartitions`. This function could return a wrong result when the partitioning column names are wrong. For example, if we pass `a='0'`, it will return all the partitions. The expected return should be an empty set, right?

Two possible solutions:

1. get the whole list, and filter it out by ourselves. Do you want me to write a test case and fix this issue?
2. check if the specs contains any column that is not part of table partitioning at the beginning.  @andrewor14 @yhuai

gatorsmile replied:
Hive does not have such an issue. It detects the error and reports
```
hive> ALTER TABLE extTable_with_partitions DROP PARTITION (a='0');
FAILED: SemanticException Column a not found
```

yhuai replied:
I think SessionCatalog should take care the check.

yhuai replied:
It's better to avoid of fetching all partition specs to the client side. If there is a large table, fetching all partition spec will be very slow.

yhuai replied:
You meant that `a` was not a partitioning column, right?

gatorsmile replied:
yeah. True

gatorsmile replied:
Will do it soon. Thanks!

yhuai replied:
https://issues.apache.org/jira/browse/SPARK-14603 is the JIRA for the work of using SessionCatalog to check if a metadata operation is valid or not.

gatorsmile replied:
Yeah, will use this JIRA to handle all the cases. Thanks!
